cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs

```diff
@@ -0,0 +1,350 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.CLI.Tx.BenchmarkingTxSubmission
+  ( ROEnv(..)
+  , RPCTxSubmission(..)
+  , TraceBenchTxSubmit(..)
+  , bulkSubmission
+  )
+where
+
+import           Prelude
+import           Cardano.Prelude (MonadIO, Word16, Seq, StateT
+                                 , evalStateT, gets, isJust, lift, {- liftIO, -} modify, toList, unless, when)
+
+import           Control.Exception (assert)
+import           Control.Monad.Class.MonadSTM ( MonadSTM, LazyTMVar, LazyTVar
+                                              , atomically, putTMVar, readTVar, retry, takeTMVar
+                                              , tryTakeTMVar)
+import           Control.Monad.Class.MonadTime ( MonadTime(..)
+                                               , addTime, diffTime)
+import           Control.Monad.Class.MonadTimer (MonadTimer, threadDelay)
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.Sequence as Seq
+import qualified Data.Set as Set
+import           Data.Time.Clock (DiffTime)
+import qualified Ouroboros.Consensus.Mempool as Mempool
+import qualified Ouroboros.Network.Protocol.TxSubmission.Type as TxSubmit
+
+-- | Bulk submisson of transactions.
+--
+--   This is intended to be used as a seperate process that interfaces
+--   between some form of transaction generation mechanism
+--   (e.g. benchmarking) and the local Ouroboros protocol peer. It has
+--   configurable latency hiding (number of transactions to try to
+--   keep ahead of actual submission) and has tracing that both
+--   reports observables related to transaction submission and whether
+--   the tx generation is able to keep up the pace.
+--
+--   For the details of the protocol description see
+--   `TxSubmit.TxSubmission` and its associated instances
+--   (e.g. `Message`).
+
+bulkSubmission :: forall m blk txid tx .
```

Comment:
It's better to move the `:: forall m blk txid tx .` to the next line and reindent -- so that the offset of the entire block doesn't depend on the length of the `bulkSubmission` name.